### PR TITLE
Add builtin function load_module

### PR DESCRIPTION
**Pull request by mfontani**
As per PPC006[1]

[1]: https://github.com/Perl/PPCs/blob/main/ppcs/ppc0006-load-module.md

### CODE REVIEW

File: lib/builtin.t
Comment (leonerd) [inline lib/builtin.t]: No need to string-eval this one. For simply capturing exceptions, you probably want to `eval { ... }` or even consider using the `try` feature.

And the same for the remaining cases in the test file.

File: lib/builtin.pm
Comment (leonerd) [inline lib/builtin.pm]: No need for the particle here; just begin the sentence "Loads a named module ..."

File: lib/builtin.pm
Comment (leonerd) [inline lib/builtin.pm]: I'd word this more, "does not yield a true value at the end of its scope". Plus maybe a mention of the new `yield_true` feature of `use v5.38`.

File: builtin.c
Comment (leonerd) [inline builtin.c]: My apologies - I think I sent you down the wrong path in my original email.

Having looked more into this since then I've remembered we have a `load_module` function at the C level as well. That would be a far simpler way to do all of this.

```c
load_module(PERL_LOADMOD_NOIMPORT, module_name, NULL, NULL);
```

This is the one I was thinking of when I mentioned `utilize()`. Sorry for the misdirection.

File: lib/builtin.pm
Comment (leonerd) [inline lib/builtin.pm:377]: As it's new (and therefore experimental) don't forget to copy the "This function is experimental" text from one of the other function docs.

**Review by leonerd — CHANGES_REQUESTED** [review]

File: lib/builtin.t
Comment (Grinnz) [reply lib/builtin.t]: Seems somewhat fragile to rely on the system not having a Foo module to load, but maybe that is acceptable for core tests?

Comment (demerphq) [reply lib/builtin.t]: It should just load something we have that dies during load. Eg, t/lib/Dies.pm

**Comment by mfontani** [comment]
@leonerd looks like this should be all, please have another look.

File: lib/builtin.pm
Comment (mfontani) [reply lib/builtin.pm:377]: Done that in latest commit.

File: builtin.c
Comment (Leont) [inline builtin.c]: This line seems unnecessary, why not go straight for `SV *module_name = newSVsv(ST(0));`?

Comment (mfontani) [reply builtin.c]: Fair point. Rebased and squashed, thanks.

File: lib/builtin.pm
Comment (leonerd) [inline lib/builtin.pm]: Typo "evaluate"

File: lib/builtin.t
Comment (leonerd) [inline lib/builtin.t:609]: Another test worth adding is the usual magic one; something like:

```perl
"Dies" =~ m/(\w+)/;
load_module($1);
...
```

and then check that it still behaves correctly. I believe your implementation currently won't, but it can be fixed by adding a call to `SvGETMAGIC(ST(0));` on a line before the `newSVsv`.

**Review by leonerd — CHANGES_REQUESTED** [review]

File: lib/builtin.pm
Comment (mfontani) [reply lib/builtin.pm]: Well spotted. Updated.

File: lib/builtin.t
Comment (mfontani) [reply lib/builtin.t:609]: I've added four of the type `load_module($1)`, two for failure and two for success: `$module_name =~ m!(\w+)!` and `"..." =~ m!(\w+)!` and all four seem to pass (on my machine) without needing to `SvGETMAGIC(...)`.  I'll watch test suite failures in case I can find other ways to reliably repro the requirement for having to add that call.

Comment (mfontani) [reply lib/builtin.t:609]: Looks like those passed everywhere despite the lack of `SvGETMAGIC(..)`. Would you know why it should in principle be required, or have a suggestion for a test what would highlight it being required? Looks otherwise to not be required...

Comment (leonerd) [reply lib/builtin.t:609]: Ooohright; yes. Because you're doing `newSVsv()` that already invokes get magic on its source argument.

In that case the code is fine but remember to add those tests in case of future edits.

Comment (mfontani) [reply lib/builtin.t:609]: Tests for that are there:
- https://github.com/Perl/perl5/pull/21532/files#diff-3b3171c64e546f883b025fda758c78654dc5e73380370bf557acd6d9a10f7991R472-R489 for the failure to load a module (`$module_name = "Dies"` vs "Dies", then perform `=~` on either)
- https://github.com/Perl/perl5/pull/21532/files#diff-3b3171c64e546f883b025fda758c78654dc5e73380370bf557acd6d9a10f7991R507-R522 for the success (ditto, for `strict` instead of `Dies`

**Comment by haarg** [comment]
This appears to segfault when given unexpected input, like undef or a reference.

Also, the `Perl_load_module` being used internally uses a `BEGIN` block. This isolates some bits of hints etc and impacts the caller stack. I wouldn't say that is specifically bad, but it is a difference from `require`, which the documentation implies shouldn't be the case.

The return value from this function is not tested, but seems to just be the same as its input module name. This is not obviously intentional, and not documented. The PPC this is meant to implement also doesn't state what the return value should be.

**Comment by leonerd** [comment]
Ahyes; PPC0006 doesn't talk about the return value. I'd suggest it should either be literally nothing, or just `PL_sv_yes`.

**Comment by mfontani** [comment]
Thanks @haarg & @leonerd, updated:

- croak if `!SVpOK(...)` so it croaks on literal undef or reference, etc & add tests for that
- return `XSRETURN_YES` (i.e. "a true value"), document that and test that
- add a test for tied scalar, and for `load_module()` failing to compile (as string eval...)

Not handled, "really":

> Also, the Perl_load_module being used internally uses a BEGIN block. This isolates some bits of hints etc and impacts the caller stack. I wouldn't say that is specifically bad, but it is a difference from require, which the documentation implies shouldn't be the case.

I'm unsure how to be able to both call the `load_module` _and_ **not** make it do that... so I've amended the documentation stating it's pretty much the same as `require` _but_ it "internally uses a `C<BEGIN>` block to perform its operations".  The difference is subtle, and there's corner cases, but it's _mostly_ like require.

Let me know your thoughts.

File: builtin.c
Comment (tonycoz) [inline builtin.c]: The call to load_module() will release `module_name`, but `module_name` isn't released if the croak is called.

Comment (mfontani) [reply builtin.c]: Thanks! I've added a `SvREFCNT_dec(module_name);` prior to calling `croak_xs_usage(cv, "defined string");`. That should do it, right?

**Review by tonycoz — APPROVED** [review]

**Comment by mfontani** [comment]
Is there anything else I should be doing for this?

File: lib/builtin.pm
Comment (leonerd) [inline lib/builtin.pm]: Probably best not to explicitly document that the `BEGIN` block is used, in case we change the implementation at another time. Keeping it somewhat vague gives us the freedom to change it in future.

Moreover probably better to downplay the "it's like require but..." in favour of simply describing what it does in its own terms. At the end you could say that it's a bit similar to require, without specifically detailing exactly the differences.

**Review by leonerd — COMMENTED** [review]
I think this is getting close to ready now.

We discussed this at the PSC meeting today. I've put some comments about the docs above.

We further thought about what the return value ought to be. The original PPC document doesn't discuss it. The implementation you've gone for is `return true`, which seems fine. But also a valid choice could be returning the module name itself. It would permit things like

```perl
my @objects = map { load_module($_)->new } @class_names;

BEGIN { load_module("Some::Thing")->import(qw( some symbol names )); }
```

Neither of these is particularly difficult to do as it stands, but that might be an extra convenience. Either way, probably not worth blocking this PR for that. If we decide to change it we can alter it later down the line.

File: lib/builtin.pm
Comment (mfontani) [reply lib/builtin.pm]: Thanks. Fair point. See if the updated version - which removes that part - is better.

It merely states: "the effect of ... is _mostly the same as_ requiring it".

**Comment by mfontani** [comment]
> But also a valid choice could be returning the module name itself. It would permit things like
> 
> ```perl
> my @objects = map { load_module($_)->new } @class_names;
> 
> BEGIN { load_module("Some::Thing")->import(qw( some symbol names )); }
> ```

I had that initially. Pushed a separate commit which does that.

If approved, I'd like to squash them though.

**Comment by mfontani** [comment]
Added a useful test for `load_module(...)->import(...)`.

**Comment by mfontani** [comment]
Re-pushed (and squashed) after conflicts were found in `pod/perldelta.pod`.

File: lib/builtin.pm
Comment (leonerd) [inline lib/builtin.pm]: Since the implementation and tests now work to the idea of returning the module name, these docs need fixing.

File: builtin.c
Comment (leonerd) [inline builtin.c:492]: I'd add a comment here specifically pointing out we leave the module name on the stack, so that becomes its return value. It's intentional. Also probably good to put an `XSRETURN(1)` to ensure we return exactly one item even in list context.

**Review by leonerd — CHANGES_REQUESTED** [review]
Generally good, a couple more little comments need adding.

File: lib/builtin.pm
Comment (mfontani) [reply lib/builtin.pm]: Thanks, updated with:

> Returns the loaded module's name on success.

File: builtin.c
Comment (mfontani) [reply builtin.c:492]: Thanks! Updated with:

```c
    /* The loaded module's name is left intentionally on the stack for the
     * caller's benefit, and becomes load_module's return value. */
    XSRETURN(1);
```

**Review by leonerd — APPROVED** [review]
LGTM

If there's no other objections to this over the wekeend, I say we merge it on Monday.

**Comment by Leont** [comment]
[PerlIO::import](https://github.com/Perl/perl5/blob/blead/lib/PerlIO.pm#L22) may be a good first user. It's core-only so is free to use such a feature.

**Comment by mfontani** [comment]
> [PerlIO::import](https://github.com/Perl/perl5/blob/blead/lib/PerlIO.pm#L22) may be a good first user. It's core-only so is free to use such a feature.

Sounds like a good idea. I will get on it once this is merged, though, and not as part of this PR.

**Comment by iabyn** [comment]
I think the POD for builtin::load_module should state (briefly) its raison d'etre, i.e. what problem does it solve which require can't do.  Which is that require doesn't accept module names (apart from barewords) - it only accepts filenames.

**Pull request merged**
